### PR TITLE
project files, locations and other misc changes

### PR DESCRIPTION
**Pull request by fgetcat**
* remap f5 crash button to f9 so that people dont crash saturn trying to toggle timeline using f6 (suggested by sm64mods)
* camera shake toggle
* water bomb spawner toggle
* butterfly toggle
* saveable saturn "project" files
* tiltable camera (using keys V and B)
* chromakey color can be animated
* splash screen to get something to draw while loading models and stuff
* locations (aka minecraft waypoints)
* unlock doors disables infinite stairs
* lighting color

Project file limitations:
* If a level has more than 4 areas, the 5th area will warp you to the 1st one
* Camera restoration is not perfect
* Current animation, Mario speed and expressions (including current model, eye texture, etc.) doesn't get stored

### CODE REVIEW

File: src/saturn/saturn_projectfile.cpp [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+    } u;
+    u.x = data[*offset];
+    *offset += 1;
+    return u.y;
+}
+uint get_int32(char* data, int* offset) {
+    uint value = (get_int8(data, offset) << 24) | (get_int8(data, offset) << 16) | (get_int8(data, offset) << 8) | get_int8(data, offset);
+    return value;
+}
+ushort get_int16(char* data, int* offset) {
```
Comment (integerbang) [inline src/saturn/saturn_projectfile.cpp]: `ushort` should be either defined or changed to `u16` or `uint16_t`.

File: src/saturn/saturn_projectfile.cpp [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+    ushort value = (get_int8(data, offset) << 8) | get_int8(data, offset);
```
Comment (integerbang) [inline src/saturn/saturn_projectfile.cpp]: Same issue as line [100](https://github.com/Llennpie/Saturn/blob/f987112d8bf0845917a5dc8266ab123e64f72400/src/saturn/saturn_projectfile.cpp#L100).

File: src/saturn/saturn_projectfile.cpp [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+}
+
+void put_int32(char* data, int* offset, uint value) {
+    data[*offset] = (value >> 24) & 0xFF;
+    data[*offset + 1] = (value >> 16) & 0xFF;
+    data[*offset + 2] = (value >> 8) & 0xFF;
+    data[*offset + 3] = value & 0xFF;
+    *offset += 4;
+}
+void put_int16(char* data, int* offset, ushort value) {
```
Comment (integerbang) [inline src/saturn/saturn_projectfile.cpp]: Same issue as lines [100](https://github.com/Llennpie/Saturn/blob/f987112d8bf0845917a5dc8266ab123e64f72400/src/saturn/saturn_projectfile.cpp#L100) and [101](https://github.com/Llennpie/Saturn/blob/f987112d8bf0845917a5dc8266ab123e64f72400/src/saturn/saturn_projectfile.cpp#L101).

File: src/saturn/saturn_projectfile.cpp [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+    k_frame_keys.clear();
+    while (true) {
+        int beginning = pointer;
+        char sectionIdentifier[5];
+        identifier(data, &pointer, sectionIdentifier);
+        uint size = get_int32(data, &pointer);
+        align(&pointer, SATURN_PROJECT_BLOCK_SIZE);
+        if (std::strcmp(sectionIdentifier, SATURN_PROJECT_DONE_IDENTIFIER) == 0) break;
+        if (std::strcmp(sectionIdentifier, SATURN_PROJECT_GAME_IDENTIFIER) == 0) {
+            ushort flags = get_int16(data, &pointer);
```
Comment (integerbang) [inline src/saturn/saturn_projectfile.cpp]: Same issue as lines [100](https://github.com/Llennpie/Saturn/blob/f987112d8bf0845917a5dc8266ab123e64f72400/src/saturn/saturn_projectfile.cpp#L100), [101](https://github.com/Llennpie/Saturn/blob/f987112d8bf0845917a5dc8266ab123e64f72400/src/saturn/saturn_projectfile.cpp#L101), and [138](https://github.com/Llennpie/Saturn/blob/f987112d8bf0845917a5dc8266ab123e64f72400/src/saturn/saturn_projectfile.cpp#L138).

**Review by integerbang — CHANGES_REQUESTED** [review]

**Comment by integerbang** [comment]
@Dominicentek, sorry for not being around to review your changes quickly, by the way.

I'm a bit busy with some other stuff. Once I have time to review your changes, I'll let you know if they're all okay for merging or if there are still some improvements that could be made. 👍

**Comment by integerbang** [comment]
I should mention that it might be better to split some of your changes into a separate pull request since I see there's a lot of new functionality to look over as a result of your changes.

For example, some things not wholly related to the new project file format could be split into their own pull request, while most of the stuff involving corrections or simple bug fixes could be merged with the other commits.

**Comment by fgetcat** [comment]
Ah okay, I understand

**Review by integerbang — APPROVED** [review]
Waiting for @Llennpie to merge or review.

**Pull request merged**